src/h3lib/lib/algos.c

```diff
@@ -677,36 +673,162 @@ void H3_EXPORT(polyfill)(const GeoPolygon* geoPolygon, int res, H3Index* out) {
+        Geofence* hole = &(geoPolygon->holes[i]);
+        for (int j = 0; j < hole->numVerts; j++) {
+            H3Index pointHex = H3_EXPORT(geoToH3)(&(hole->verts[j]), res);
+            // A simple hash to store the hexagon, or move to another place if
+            // needed
+            int loc = (int)(pointHex % numHexagons);
+            int loopCount = 0;
+            while (found[loc] != 0) {
+                if (loopCount > numHexagons) {
+                    free(search);
+                    free(found);
+                    free(bboxes);
+                    return -1;
+                }
+                if (found[loc] == pointHex)
+                    break;  // At least two points of the geofence are identical
+                loc = (loc + 1) % numHexagons;
+                loopCount++;
+            }
+            if (found[loc] == pointHex)
+                continue;  // Skip this hex, already exists in the found hash
+            // Otherwise, set it in the found hash for now
+            found[loc] = pointHex;
+
+            // Set the hexagon in the search hash, as well
+            search[numSearchHexes] = pointHex;
+            numSearchHexes++;
         }
-        // Check if hexagon is inside of polygon
-        GeoCoord hexCenter;
-        H3_EXPORT(h3ToGeo)(out[i], &hexCenter);
-        hexCenter.lat = constrainLat(hexCenter.lat);
-        hexCenter.lon = constrainLng(hexCenter.lon);
-        // And remove from list if not
-        if (!pointInsidePolygon(geoPolygon, bboxes, &hexCenter)) {
-            out[i] = H3_INVALID_INDEX;
+    }
+
+    // 3. Re-zero the found hash so it can be used in the main loop below
+    for (int i = 0; i < numHexagons; i++) found[i] = 0;
+
+    // 4. Begin main loop. While the search hash is not empty do the following
+    while (numSearchHexes > 0) {
+        // Iterate through all hexagons in the current search hash, then loop
+        // through all neighbors and test Point-in-Poly, if point-in-poly
+        // succeeds, add to out and found hashes if not already there.
+        H3Index ring[7] = {0};
+        int currentSearchNum = 0;
+        int i = 0;
+        while (currentSearchNum < numSearchHexes) {
+            H3Index searchHex = search[i];
+            if (searchHex == 0) {  // Gap, skip it
+                i++;
+                continue;
+            }
+            H3_EXPORT(kRing)(searchHex, 1, ring);
+            for (int j = 1; j < 7; j++) {
+                if (ring[j] == 0) {
+                    continue;  // Skip if this was a pentagon and only had 5
+                               // neighbors
+                }
+
+                H3Index hex = ring[j];
+
+                // A simple hash to store the hexagon, or move to another place
+                // if needed. This MUST be done before the point-in-poly check
+                // since that's far more expensive
+                int loc = (int)(hex % numHexagons);
```

Comment:
I can test the function approach, but this is sitting inside of the core of the loop, so I figured it'd be better to avoid the overhead when possible.